hash-table/chaining.js

```diff
@@ -0,0 +1,139 @@
+const { Node, DoublyLinkedList } = require('../linked-list/doubly-linked-list')
+
+class ChainingHashTable {
+  constructor(capacity = 8) {
+    this.arr = new Array(capacity)
+    this.capacity = capacity
+    this.size = 0
+  }
+
+  hash(key) {
+    return key % this.capacity
+  }
+
+  insert(value, key) {
+    const index = this.hash(key)
+    if (!this.arr[index]) {
+      const linkList = new DoublyLinkedList()
+      linkList.pushFront(value, key)
+      this.arr[index] = linkList
+      this.size++
+    } else {
+      const isExisting = this.search(key)
+      if (isExisting) {
+        let cur = this.arr[index].tail
+        while (cur) {
+          if (cur.key === key) {
+            cur.value = value
+            break
+          }
+          cur = cur.prev
+        }
+      } else {
+        this.arr[index].pushBack(value, key)
+      }
+    }
+
+    if (this.size === this.capacity) {
+      this.extendCapacity()
+    }
+  }
+
+  delete(key) {
+    const index = this.hash(key)
+    if (this.arr[index] && this.arr[index].length > 1) {
+      let cur = this.arr[index].tail
+      let count = 0
+      
+      while (cur) {
+        if (cur.value.value === key) {
+          let len = this.arr[index].length
+          this.arr[index].removeAt(len - count - 1)
+          break
+        }
+        count++
+        cur = cur.prev
+      }
+    } else {
+      this.arr[index] = undefined
+      this.size--
+    }
+    if (this.capacity > 8 && this.size <= Math.floor(this.capacity / 4)) {
```

Comment:
We don't keep the magic numbers in the code, Mr @trungducc said :)) Grow/shrink factor, default capacity, etc, all should be declared as constants somewhere easy to change, so we don't have to go to each function and change them.